pkg/functionconfig/types.go

```diff
@@ -568,6 +572,89 @@ func (c *Config) EnrichWithTolerations(tolerations []v1.Toleration) {
 	}
 }
 
+// PruneAffinityNodeSelectorRequirement prunes given affinity node selector requirement from function spec
+func (c *Config) PruneAffinityNodeSelectorRequirement(nodeSelectorRequirements []v1.NodeSelectorRequirement) {
+
+	// nothing to do here
+	if c.Spec.Affinity == nil || nodeSelectorRequirements == nil {
+		return
+	}
+
+	if nodeAffinity := c.Spec.Affinity.NodeAffinity; nodeAffinity != nil {
+		if nodeSelector := nodeAffinity.RequiredDuringSchedulingIgnoredDuringExecution; nodeSelector != nil {
+
+			// for all term expressions on function spec
+			for termIdx, term := range nodeSelector.DeepCopy().NodeSelectorTerms {
+				for expressionIndex, expression := range term.MatchExpressions {
+
+					// check if its key matches the anti affinity
+					// if it does, we want to remove this expression so it wont block us
+					for _, nodeSelectorRequirement := range nodeSelectorRequirements {
+						if nodeSelectorRequirement.Key == expression.Key &&
+							nodeSelectorRequirement.Operator == expression.Operator &&
+							reflect.DeepEqual(nodeSelectorRequirement.Values, expression.Values) {
+
+							// slice expressions
+							nodeSelector.NodeSelectorTerms[termIdx].MatchExpressions = append( // nolint: gocritic
+								term.MatchExpressions[:expressionIndex],
+								term.MatchExpressions[expressionIndex+1:]...,
+							)
+							break
+						}
+					}
+				}
+
+				if len(nodeSelector.NodeSelectorTerms[termIdx].MatchExpressions) == 0 &&
+					len(nodeSelector.NodeSelectorTerms[termIdx].MatchFields) == 0 {
+
+					// slice term
+					nodeSelector.NodeSelectorTerms = append(
+						nodeSelector.NodeSelectorTerms[:termIdx],
+						nodeSelector.NodeSelectorTerms[termIdx+1:]...)
+				}
```

Comment:
I kinda liked the "iterating on a copy to be able to update the origin" idea until here where you starting slicing indexes
I feel like it will be more readable to simply build a new list when iterating (skipping the ones you want to remove) and assign after the loop